arch/arc/core/fault_s.S

```diff
@@ -109,15 +109,47 @@ exc_nest_handle:
 
 #ifdef CONFIG_IRQ_OFFLOAD
 GTEXT(_irq_do_offload);
+#endif
 
+#if defined(CONFIG_IRQ_OFFLOAD) || defined(CONFIG_USERSPACE)
 SECTION_SUBSEC_FUNC(TEXT,__fault,__ev_trap)
+#ifdef CONFIG_USERSPACE
+	/* get the id of trap_s */
+	lr ilink, [_ARC_V2_ECR]
+	and ilink, ilink, 0x3f
+	cmp ilink, 0x3
+	bne _do_other_trap
+/* do sys_call */
+	mov ilink, _SYSCALL_LIMIT
+	cmp r6, ilink
+	blt valid_syscall_id
+
+	mov r0, r6
+	mov r6, _SYSCALL_BAD
+
+valid_syscall_id:
+	lr ilink, [_ARC_V2_ERET]
+	push ilink
+	lr ilink, [_ARC_V2_ERSTATUS]
+	push ilink
+
+	bclr ilink, ilink, _ARC_V2_STATUS32_U_BIT
+	sr ilink, [_ARC_V2_ERSTATUS]
+
+	mov ilink, _arc_do_syscall
+	sr ilink, [_ARC_V2_ERET]
+
+	rtie
```

Comment:
I'm not seeing any mechanism to scrub registers of kernel context before returning to userspace.
Does the hardware automatically do this?

Reply:
No, hardware isn't doing that (except for user/kernel stack pointers). Relevant registers would be all general purpose registers, except the ones used for syscall parameter passing, right?

Reply:
basically we can't have any register state due to the kernel processing the system call leak back to user mode when we transition back as it may contain sensitive information that only the kernel should know about.

Reply:
It's better to clean all the regs when returns to user mode.  But there is a performance concern, ARC has a lot of gp regs, it will require extra 20-30 instructions to clear this regs.

Reply:
Unfortunately we need to do this for people who are using memory protection as a security feature.
If you really feel it's a performance killer, make it a Kconfig which defaults to on.

Reply:
I already add the cleanup of regs in the latest commits